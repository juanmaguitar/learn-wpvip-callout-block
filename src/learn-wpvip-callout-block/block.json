{
	"$schema": "https://schemas.wp.org/trunk/block.json",
	"apiVersion": 2,
	"name": "learn-wpvip/callout",
	"version": "0.1.0",
	"title": "Callout Block",
	"category": "design",
	"icon": "marker",
	"description": "Add a color-coded callout to your post.",
	"supports": {
		"align": true,
		"spacing": {
			"margin": ["top", "bottom"],
			"padding": true,
			"blockGap": false
		},
		"typography": {
			"fontSize": true,
			"lineHeight": true
		}
	},
	"attributes": {
		"content": {
			"type": "string",
			"source": "html",
			"selector": ".wp-block-learn-wpvip-callout__content",
			"multiline": "p"
		},
		"type": {
			"type": "string",
			"default": "tip",
			"enum": ["alert", "info", "warning", "tip", "success"]
		}
	},
	"example": {
		"attributes": {
			"type": "alert",
			"content": "<p>This is an alert callout.</p>"
		}
	},
	"textdomain": "learn-wpvip",
	"editorScript": "file:./index.js",
	"style": "file:./style-index.css"
}
