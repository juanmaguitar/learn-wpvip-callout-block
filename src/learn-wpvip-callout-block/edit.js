/**
 * WordPress dependencies
 */
import { __, sprintf } from "@wordpress/i18n";
import {
	BlockControls,
	RichText,
	useBlockProps,
} from "@wordpress/block-editor";
import { ToolbarDropdownMenu } from "@wordpress/components";

/**
 * Internal dependencies
 */
import Icon from "./icon";

export const typeOptions = [
	{ label: __("Tip", "learn-wpvip"), value: "tip" },
	{ label: __("Success", "learn-wpvip"), value: "success" },
	{ label: __("Information", "learn-wpvip"), value: "info" },
	{ label: __("Alert", "learn-wpvip"), value: "alert" },
	{ label: __("Warning", "learn-wpvip"), value: "warning" },
];

/**
 * Get the human-readable label from the current type value.
 *
 * @param {string} type The type slug.
 * @return {string} The translated human-friendly label.
 */
function getOptionLabel(type) {
	const currentType = typeOptions.find(({ value }) => type === value);
	return currentType.label || "";
}

export function Edit({ attributes, setAttributes }) {
	const { content, type } = attributes;
	const className = `is-${type}-callout`;

	return (
		<>
			<BlockControls group="block">
				<ToolbarDropdownMenu
					text={sprintf(
						/* translators: %s is the notice type label. */
						__("Type: %s", "learn-wpvip"),
						getOptionLabel(type),
					)}
					icon={false}
					controls={typeOptions.map(({ label, value }) => ({
						title: label,
						icon: <Icon type={value} />,
						isActive: value === type,
						onClick: () => setAttributes({ type: value }),
					}))}
				/>
			</BlockControls>
			<div {...useBlockProps({ className })}>
				<div className="wp-block-learn-wpvip-callout__icon" />
				<RichText
					tagName="div"
					multiline="p"
					className="wp-block-learn-wpvip-callout__content"
					onChange={(newContent) => setAttributes({ content: newContent })}
					value={content}
				/>
			</div>
		</>
	);
}
