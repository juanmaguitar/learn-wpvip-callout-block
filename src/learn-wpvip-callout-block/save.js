/**
 * WordPress dependencies
 */
import { RichText, useBlockProps } from "@wordpress/block-editor";

export default function save({ attributes }) {
	const { content, type } = attributes;
	const className = `is-${type}-callout`;

	return (
		<div {...useBlockProps.save({ className })}>
			<div className="wp-block-learn-wpvip-callout__icon" />
			<div className="wp-block-learn-wpvip-callout__content">
				<RichText.Content multiline="p" value={content} />
			</div>
		</div>
	);
}
