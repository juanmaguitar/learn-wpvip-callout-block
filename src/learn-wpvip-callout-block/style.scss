/**
 * The following styles get applied both on the front of your site
 * and in the editor.
 *
 * Replace them with your own styles or remove the file completely.
 */

.wp-block-learn-wpvip-callout {
	--wp--custom--vip-learn-callout--color--background: var(
		--wp--custom-support-background-info,
		rgba(205, 230, 235, 1)
	);

	--wp--style--global--content-size: 680px;

	&.is-tip-callout {
		--wp--custom--vip-learn-callout--color--background: var(
			--wp--custom-support-background-success,
			rgba(198, 234, 208, 1)
		);
	}

	&.is-success-callout {
		--wp--custom--vip-learn-callout--color--background: var(
			--wp--custom-support-background-success,
			rgba(198, 234, 208, 1)
		);
	}

	&.is-info-callout {
		--wp--custom--vip-learn-callout--color--background: var(
			--wp--custom-support-background-info,
			rgba(205, 230, 235, 1)
		);
	}

	&.is-alert-callout {
		--wp--custom--vip-learn-callout--color--background: var(
			--wp--custom-support-background-error,
			rgba(255, 213, 206, 1)
		);
	}

	&.is-warning-callout {
		--wp--custom--vip-learn-callout--color--background: var(
			--wp--custom-support-background-warning,
			rgba(255, 216, 165, 1)
		);
	}

	display: grid;
	grid-template-columns: auto 1fr;
	gap: 0.5em;
	padding: 1.25em;
	margin-bottom: 1.25em;
	background-color: var(--wp--custom--vip-learn-callout--color--background);
	border-radius: 2px;

	& > * {
		align-self: start;
	}

	&.alignleft,
	&.alignright {
		max-width: calc(var(--wp--style--global--content-size, 680px) * 0.66);
	}
}

.wp-block-learn-wpvip-callout__icon {
	display: flex;
	align-self: flex-start;

	/* Default line height for the content, to optically align icon + first line of content. */
	height: 1.875em;
	width: 24px;

	background-image: url(./icon/library/tip.svg);
	background-repeat: no-repeat;
	background-position: top;
	background-size: 24px 24px;
	margin-top: 1px;
}

.wp-block-learn-wpvip-callout__content {
	& > *:last-child {
		padding-bottom: 0;
	}
}

.is-success-callout .wp-block-learn-wpvip-callout__icon {
	background-image: url(./icon/library/success.svg);
}

.is-info-callout .wp-block-learn-wpvip-callout__icon {
	background-image: url(./icon/library/info.svg);
}

.is-alert-callout .wp-block-learn-wpvip-callout__icon {
	background-image: url(./icon/library/alert.svg);
}

.is-warning-callout .wp-block-learn-wpvip-callout__icon {
	background-image: url(./icon/library/warning.svg);
}
