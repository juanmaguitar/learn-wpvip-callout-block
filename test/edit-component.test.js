import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { Edit } from "../src/learn-wpvip-callout-block/edit";

const mockEditProps = {
	attributes: {
		type: "tip",
		content: "<p>Test content</p>",
	},
	setAttributes: jest.fn(),
	clientId: "test-id",
	className: "wp-block-learn-wpvip-callout",
};

beforeEach(() => {
	mockEditProps.setAttributes.mockClear();
	jest.clearAllMocks();
});

describe("Edit Component", () => {
	it("matches snapshot", () => {
		const { container: renderedEditComponent } = render(
			<Edit {...mockEditProps} />,
		);
		expect(renderedEditComponent).toMatchSnapshot();
	});

	describe("Type Selection", () => {
		it("updates type when changed", async () => {
			// Setup
			const { setAttributes } = mockEditProps;
			const user = userEvent.setup();

			// Render
			render(<Edit {...mockEditProps} />);

			// Verify initial state
			expect(screen.getByTestId("block-controls")).toBeInTheDocument();
			expect(screen.getByTestId("toolbar-dropdown")).toBeInTheDocument();
			expect(screen.getByTestId("rich-text")).toBeInTheDocument();

			// Action
			const alertButton = screen.getByTestId("control-Alert");
			await user.click(alertButton);

			// Assertion
			expect(setAttributes).toHaveBeenCalledTimes(1);
			expect(setAttributes).toHaveBeenCalledWith({
				type: "alert",
			});
		});
	});
});
