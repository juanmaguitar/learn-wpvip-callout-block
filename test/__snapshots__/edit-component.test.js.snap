// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Component matches snapshot 1`] = `
<div>
  <div
    data-testid="block-controls"
  >
    <div
      data-testid="toolbar-dropdown"
    >
      <button
        data-testid="control-Tip"
      >
        Tip
      </button>
      <button
        data-testid="control-Success"
      >
        Success
      </button>
      <button
        data-testid="control-Information"
      >
        Information
      </button>
      <button
        data-testid="control-Alert"
      >
        Alert
      </button>
      <button
        data-testid="control-Warning"
      >
        Warning
      </button>
    </div>
  </div>
  <div
    class="wp-block-learn-wpvip-callout is-tip-callout"
  >
    <div
      class="wp-block-learn-wpvip-callout__icon"
    />
    <div
      class="wp-block-learn-wpvip-callout__content"
      data-testid="rich-text"
    >
      <p>
        Test content
      </p>
    </div>
  </div>
</div>
`;
